From: Bhavesh Davda <bhavesh@vmware.com>
Date: Sat, 24 Jul 2010 14:43:29 +0000
Subject: [PATCH] net-next: Fix an overflow bug in vmxnet3 Tx descriptor
Patch-mainline: 2.6.36-rc2
References: bnc#624020
Git-commit: 1f4b16128439b225c2986f06d015c848c290d7d9

Fix an overflow bug in vmxnet3 Tx descriptor

This patch fixes a bug where a 16K buffer on a Tx descriptor was overflowing
into the 'gen' bit in the descriptor thereby corrupting the descriptor and
stalling the transmit ring.

Signed-off-by: Bhavesh Davda <bhavesh@vmware.com>
Signed-off-by: Shreyas Bhatewara <sbhatewara@vmware.com>
Signed-off-by: Matthew Delco <delcoM@vmware.com>
Signed-off-by: Ronghua Zhang <ronghua@vmware.com>
Signed-off-by: David S. Miller <davem@davemloft.net>
Signed-off-by: Brandon Philips <bphilips@suse.de>

---
 drivers/net/vmxnet3/vmxnet3_drv.c |   13 +++++++++----
 drivers/net/vmxnet3/vmxnet3_int.h |    4 ++--
 2 files changed, 11 insertions(+), 6 deletions(-)

Index: linux-2.6.32-SLE11-SP1/drivers/net/vmxnet3/vmxnet3_drv.c
===================================================================
--- linux-2.6.32-SLE11-SP1.orig/drivers/net/vmxnet3/vmxnet3_drv.c
+++ linux-2.6.32-SLE11-SP1/drivers/net/vmxnet3/vmxnet3_drv.c
@@ -562,8 +562,13 @@ vmxnet3_map_pkt(struct sk_buff *skb, str
 	while (len) {
 		u32 buf_size;
 
-		buf_size = len > VMXNET3_MAX_TX_BUF_SIZE ?
-			   VMXNET3_MAX_TX_BUF_SIZE : len;
+		if (len < VMXNET3_MAX_TX_BUF_SIZE) {
+			buf_size = len;
+			dw2 |= len;
+		} else {
+			buf_size = VMXNET3_MAX_TX_BUF_SIZE;
+			/* spec says that for TxDesc.len, 0 == 2^14 */
+		}
 
 		tbi = tq->buf_info + tq->tx_ring.next2fill;
 		tbi->map_type = VMXNET3_MAP_SINGLE;
@@ -571,13 +576,13 @@ vmxnet3_map_pkt(struct sk_buff *skb, str
 				skb->data + buf_offset, buf_size,
 				PCI_DMA_TODEVICE);
 
-		tbi->len = buf_size; /* this automatically convert 2^14 to 0 */
+		tbi->len = buf_size;
 
 		gdesc = tq->tx_ring.base + tq->tx_ring.next2fill;
 		BUG_ON(gdesc->txd.gen == tq->tx_ring.gen);
 
 		gdesc->txd.addr = tbi->dma_addr;
-		gdesc->dword[2] = dw2 | buf_size;
+		gdesc->dword[2] = dw2;
 		gdesc->dword[3] = 0;
 
 		dev_dbg(&adapter->netdev->dev,
Index: linux-2.6.32-SLE11-SP1/drivers/net/vmxnet3/vmxnet3_int.h
===================================================================
--- linux-2.6.32-SLE11-SP1.orig/drivers/net/vmxnet3/vmxnet3_int.h
+++ linux-2.6.32-SLE11-SP1/drivers/net/vmxnet3/vmxnet3_int.h
@@ -73,10 +73,10 @@
 /*
  * Version numbers
  */
-#define VMXNET3_DRIVER_VERSION_STRING   "1.0.13.0-k"
+#define VMXNET3_DRIVER_VERSION_STRING   "1.0.14.0-k"
 
 /* a 32-bit int, each byte encode a verion number in VMXNET3_DRIVER_VERSION */
-#define VMXNET3_DRIVER_VERSION_NUM      0x01000B00
+#define VMXNET3_DRIVER_VERSION_NUM      0x01000E00
 
 
 /*
